### PR TITLE
WIP: robustirc support

### PR DESCRIPTION
**Pull request by stapelberg**
This is a very very rough first proof of concept implementation of how
native RobustIRC support in WeeChat could work.

The intention of this commit is to serve as a base for discussion, not
to be used or even committed to the repository.

### CODE REVIEW

File: src/plugins/irc/irc-server.c [hunk 1]
```diff
@@ -3387,7 +3439,7 @@ irc_server_connect_cb (void *data, int status, int gnutls_rc, int sock,
             server->current_ip = (ip_address) ? strdup (ip_address) : NULL;
             weechat_printf (
                 server->buffer,
-                _("%s%s: connected to %s/%d (%s)"),
+                "%s%s: connected to %s/%d (%s) lolwat",
```
Comment (Shawn-Smith) [inline src/plugins/irc/irc-server.c]: You've added this "lolwat" to the code in two spots, what is it for?

Comment (stapelberg) [inline src/plugins/irc/irc-server.c]: That was from the early stages of debugging, when I was still trying to get my change/compile/test workflow working for WeeChat. Please ignore :).

**Comment by stapelberg** [comment]
@flashcode So, what do you think about this pull request?

I’ve also tried coming up with an out-of-tree version of a RobustIRC plugin for WeeChat. The best approach I could find without modifying the WeeChat code is to add a signal hook for `irc_server_connecting`, then use `irc_server_search` and unhook `server->hook_connect` to abort the current connection attempt. I imagine (but haven’t tried) that I can call `irc_server_connect_cb()` with a file descriptor that I have opened in my plugin (e.g. a pipe).

The problem with that approach is that it depends on the irc plugin internals which are not exposed in the plugin API currently.

How do you think we could best extend the plugin API to make a custom IRC transport plugin possible?

(FWIW, in the meantime I have a working version of an out-of-tree irssi module, see https://github.com/robustirc/irssi-robustirc — I would love to also have one for WeeChat)

**Comment by flashcode** [comment]
Hi @stapelberg,

Sorry for the long time to reply, I was focused on other bugs and new features in WeeChat for the release of v1.2 (scheduled in two days).

So I've read RobustIRC protocol, discussions in the issue, and this pull request.

First some comments on this pull request, even if your intention is not to commit it as-is:

* The WeeChat "hooks" are protocol agnostic and low-level, so for me `hook_connect_robustirc` should not be there (there's no IRC protocol nor any other protocol in WeeChat core); `hook_fd` should be enough for socket communication, and then a plugin can implement layers over that.
* You use Curl in RobustIRC, this may block WeeChat sometimes, even if you use the multi interface of curl (see: http://curl.haxx.se/libcurl/c/libcurl-multi.html#BLOCKING); currently all network operations that can block are implemented with a fork in `wee-network.c`.
* The WeeChat plugins must never include or rely on code of WeeChat core; the only file that can be used by plugins is `weechat-plugin.h`; in other words that means any WeeChat plugin can be compiled without even having the sources of WeeChat (only `weechat-plugin.h` is required).
* Minor thing, I know you were not focused on that: automatic tests are broken by this pull request, because WeeChat can still be compiled with autotools (for historical reasons, but now CMake is preferred).

Now about the fact to include RobustIRC inside IRC plugin, as a separate plugin/script, or even as a separate proxy... this is the harder part, because I've not the best solution in mind yet.

Here are the pros and cons for me of each solution (the list is not exhaustive and we can discuss about each point):

#### Add RobustIRC inside IRC plugin (this pull request)

##### Pros

* Good integration with IRC plugin, just a boolean to toggle, and then you can use RobustIRC.

##### Cons

* As said in issue comments, the IRC plugin may become "bloated", with the add of Curl, HTTP, JSON and such in IRC plugin, which is very different from the original IRC protocol.
* It will not be possible to use RobustIRC with other protocols than IRC in WeeChat (anyway, can RobustIRC support other protocols? or it should be renamed :wink:).
* It will be harder to remove this feature (if we don't need it any more in future), because the code will be mixed with IRC plugin.
* There seem to have plans to integrate more or less similar solutions in future versions of IRC, so maybe we can wait a bit before integrating RobustIRC? (see #385)

#### Add RobustIRC as a separate C plugin

##### Pros

* Better separation of code, and if you don't need the feature, you can just not load the plugin.

##### Cons

* There's currently no way to add or replace a "transport" layer in WeeChat (especially when defined by another plugin), but we can discuss to add that: the way to do it, and to interact with other plugins (IRC in our case).
* Harder to fully integrate in IRC plugin as a server option: not easy to just add a boolean option to the server if the transport layer is managed by an external plugin.

#### Add RobustIRC as a script

##### Pros

* Faster to implement (if the API allows changes in transport layer, see the C plugin above).

##### Cons

* Too slow for me: running script callbacks use a lot of CPU, and this will be called very often, so not an acceptable solution for me.

#### Use RobustIRC as a proxy (external software, not in WeeChat)

##### Pros

* No changes in WeeChat (pros or cons? you may not agree with me here :smile:).
* Easy to setup, with the `.proxy` option in IRC server (and it can be added as a default proxy too with option `irc.server_default.proxy`, so no need to add it in all servers).
* Transparent for the client, this solution works with any IRC client (that at least supports proxy connection).

##### Cons

* Requires the external proxy to run somewhere (local or remote host).
* Adds an extra network layer between WeeChat and the RobustIRC server (if I understood well).
* It would be harder to use another proxy, like HTTP or SOCKS (not sure if the "RobustIRC bridge" itself supports proxy connection to the IRC servers).

--

In short:

* **IRC plugin**: very different from IRC protocol, joker for now on this solution.
* **Separate C plugin**: requires some work in API.
* **Script**: too slow (and requires some work in API).
* **Proxy**: already working, so we can live with that for now.

As a conclusion, I wonder if the proxy solution is not the best, are there really drawbacks to use RobustIRC as a proxy? It would be similar to services like Bitlbee, that you can install separately and connect on from the IRC client (even if RobustIRC is a bit low level compared to Bitlbee).

@stapelberg: it would be nice if we could discuss about all these points, or other things, on IRC (#weechat channel on freenode). I find that more convenient than this issue, especially for long discussions.

**Comment by JohnMaguire** [comment]
I know this ticket is five years old, but I wanted to respond to this bit:

> As a conclusion, I wonder if the proxy solution is not the best, are there really drawbacks to use RobustIRC as a proxy? It would be similar to services like Bitlbee, that you can install separately and connect on from the IRC client (even if RobustIRC is a bit low level compared to Bitlbee).

As an IRC user, the downside to this is that I have to learn and install separate software, and somehow daemonize it, which makes it more difficult to adopt. Integration in a Weechat plugin would allow me to try this out far more easily.
